src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,211 @@
+        if (table == null || table.length == 0) {
+            table = createTable(capacity);
+        } else if (size >= threshold) {
+            capacity = Math.min(2 * capacity, MAXIMUM_CAPACITY);
+            threshold = (int) (capacity * DEFAULT_LOAD_FACTOR);
+            size = 0;
+            Node<K, V>[] oldTable = table;
+            table = createTable(capacity);
+            for (Node<K, V> node : oldTable) {
+                if (node == null) {
+                    continue;
+                }
+                do {
+                    putVal(node.key, node.value);
+                } while ((node = node.next) != null);
+            }
+        }
+    }
+
+    private Node<K, V>[] createTable(final int tableCapacity) {
+        return (Node<K, V>[]) new Node[tableCapacity];
+    }
+
+    private void putVal(K key, V value) {
+        final int bucketIndex = getBucketIndex(key);
+        if (table[bucketIndex] == null) { //create first node in bucket
+            table[bucketIndex] = new Node<>(key, value, null);
+            size++;
+            return;
+        }
+        Node<K, V> node = table[bucketIndex];
+        Node<K, V> prevNode = null;
+        do { //update value in existing node (change node)
+            if ((node.key == key || (key != null && key.equals(node.key)))) {
+                if (prevNode == null) { // first node in bucket
+                    table[bucketIndex] = new Node<>(key, value, node.next);
+                    return;
+                }
+                prevNode.next = new Node<>(key, value, node.next);
+                return;
+            }
+            prevNode = node;
+        } while ((node = node.next) != null);
+        //add new node at the end of list
+        prevNode.next = new Node<>(key, value, null);
+        size++;
+    }
+
+    public V remove(K key) {
+        if (table == null || table.length == 0) {
+            return null;
+        }
+        final int bucketIndex = getBucketIndex(key);
+        Node<K, V> node = table[bucketIndex];
+        if (node == null) {
+            return null;
+        }
+        Node<K, V> prevNode = null;
+        do {
+            if ((node.key == key || (key != null && key.equals(node.key)))) {
+                if (prevNode == null) { // first node in bucket
+                    table[bucketIndex] = node.next;
+                    size--;
+                    return node.value;
+                }
+                prevNode.next = node.next;
+                size--;
+                return node.value;
+            }
+            prevNode = node;
+        } while ((node = node.next) != null);
+        return null;
+    }
+
+    private int getBucketIndex(Object key) {
+        return (key == null) ? 0 : Math.abs(key.hashCode()) % capacity;
+    }
+
+    private Node<K, V> getNode(Object key) {
+        final int bucketIndex = getBucketIndex(key);
+        if (this.table == null || this.table.length == 0) {
+            return null;
+        }
```

Comment:
You can avoid this smell code in many places - simply initialize your table in the constructor and it is never gonna be null